### ISSUE TITLE
Organising repo and code - converging to the OTP way

### ISSUE THREAD

**Issue by marjaimate** [ryal/ryal#20]
It seems like that Ryal as I can see and read it is more on a Phoenix app side, rather than a plug or even a simple Elixir app. This might work in the ruby world (equivalent like gem with database schemas, etc), I'm not sure if it's applicable on the Elixir / Phoenix world.

I'd propose the following as a direction, keeping #13 in mind as well:

1. Convert `ryal_core` to a simple Elixir OTP app
  - No models and db schemas, only payment processing
  - Use a [GenServer](http://elixir-lang.org/getting-started/mix-otp/genserver.html) to handle payment requests.
  - Consider multiple gen servers - per payment gateway
  - Consider supervisor strategy [simple_one_for_one](http://erlang.org/doc/design_principles/sup_princ.html#simple) to accommodate the traffic needs and internal process scaling
  - Interface with the world with maps
  - Break it into its own github repo

2. Convert `ryal` to a plug package and rename to `ryal_plug`
  - Keep phoenix / plug specific logic here, a wrapper for ryal_core
  - Keep the logic to record transactions / requests
  - Keep the interfaces as they are.

3. Create an umbrella package `ryal` later there will be other `ryal_*` Elixir apps (orders, etc)

I know this could be a big change, but I think will pay back hugely. I.e you could use ryal_core in other Elixir applications not just phoenix.

**Comment by James-Byrne** [comment]
This makes a lot sense to me. By abstracting parts of ryal into their own modules/apps we can better take advantage of the BEAM and OTP architecture. It would also make ryal far more configurable for the end user. Some users will just want to add payment processing to existing apps whereas others will be looking for a more complete solution with a web front end. By separating the two we can serve the needs of both. 

Its worth noting that this is more inline with the current direction of phoenix as well with 1.3 turning the framework into a simple elixir application.

**Comment by BenMorganIO** [comment]
### Number 1

Why no model or schemas? We may not be able to do a lot without them.

What do you mean by "interface with the world with maps"?

~I really wouldn't want to make it its own Github repo. That would mean making changes to Ryal in the future very brittle.~ Nah, let's do it. I had time to think this over.

### Number 2

I'm all for minimizing Ryal to just use Plug. I'm not sure if it would be a good idea though to break the project down further by segmenting web requests. This would mean that ryal_core would become an umbrella application with plug, logic, then schema. I'm not sure if it's a good idea to have umbrella apps inside of umbrella apps. Feels more like the "~turtles~ umbrellas all the way down!" argument.

I feel like this needs more discussion and deep planning as to how all the pieces would fit together harmoniously.

### Number 3

I'm already using the ryal package in an app :p

**Comment by James-Byrne** [comment]
Hey Ben, sorry for the late reply, Máté is currently in New York and I've been swamped this last week. I'll give my view on what Máté is trying to achieve here and let him chime in later to correct me. 

__1 - Models and Schemas__
I think the idea is not to get rid of the models and schemas just to move them away from ryal_core. That way we can create a clear boundary between the actual payment processing and persisting the application state. This potentially allows us to create a more flexible application where ryal_core can be used by the rest of ryal but could also be used just for payment processing in another application.

__1 - Interface with the world with maps__
Pretty sure he means that ryal_core will always take in a map and return a map when called.

__2 - Minimising to plug__
+1 I saw the issue yesterday.

__2 - Umbrella Projects__
I think the idea is to separate the two, so we have an umbrella app that contains ryal_plug for routing (and maybe schema? Not 100% sure what he meant on that one) and ryal_core for payment processing. I definitely agree that this bit needs a bit more planning though.

__3 - Pesky users__ 
Gosh darn them pesky users :P . But in all seriousness thats a good point and something we should take into account when looking at things like this. However given how young Ryal is and the size of this potential change it might be worth it to do it now rather than trying to later. Obviously we should keep the users changes small where possible and provide a clear upgrade path from one version to the next if we do go ahead with the restructure.

**Comment by BenMorganIO** [comment]
From the bottom up...

3 - We haven't reached 0.1 yet. We're at 0.0.1 mainly just to reserve the name on Hex. That's about it. We should be allowed to make as many breaking changes as we feel like ATM.

2 - With Spree/Solidus, we had all the parts broken out based on their type, not by their domain. It was `solidus_core`, `solidus_frontend`, `solidus_api`, and `solidus_backend`. I would much rather move by domain for Ryal at the first level. So we'd have:

- `ryal_core`
- `ryal_products`
- `ryal_shipping`
- `ryal_inventory`
- `ryal_subscriptions`

That's probably everything needed to get a working MVP of an e-commerce store up, but as we've discussed, we've got to highlight how the umbrella apps for each component is going to look. I think we need to focus on getting these guys split into their own repo ATM and then start thinking about how we want to split `ryal_core` out.

1 - I see what's trying to be achieved here. I think we should split it all out after we have a working core. This way we know exactly what's entailed. I'm not scared of the changes that'll be required at the end. We won't have (3) to worry about and there'll be tests. Plus, we don't fully know what we're dealing with yet. I want to see what a full functioning app looks like first.

Basically, what I'm thinking is:

1. Let's get Ryal Core split out to its own repo.
2. Convert Ryal to an OTP application.
3. Finish work on Ryal Core.
4. Split Ryal Core into an umbrella app.

**Comment by BenMorganIO** [comment]
We can probably move some of the Ryal Core splitting from Ryal discussion to here: https://github.com/ryal/ryal_core/issues/1

**Comment by marjaimate** [comment]
Hey, sorry about the late reply, just reading through the comments, and trying to address the latest ones.

The logical (i.e. thematic) split works for me (shipping, etc) and I get the argument about the 'too many apps / repos' as well. For now, hence we're in an OTP / Elixir app, we should be able to solve that within a single repo, with multiple apps. Later it can turned it into an [umbrella app](http://elixir-lang.org/getting-started/mix-otp/dependencies-and-umbrella-apps.html#umbrella-projects) with less effort.

My thinking is, that if you do the payment processing part in its own app (i.e. treat it as an internal API), it can benefit not just Phoenix / Plug users, but others who do more bare bones Elixir work. And when doing the processing, all you're concerned of is the interaction (requests and responses) and not the schema, which is more important for the one-for-all phoenix users.

As for the 'Interface with the world with maps', what I meant is instead of models and schemas (and proplists, etc) we could use simple Map objects, which reduces the learning curve for someone new. Instead of trying to understand schemas and models, you can get going as everything is a simple object.

**Comment by BenMorganIO** [comment]
OK cool. Let's move forward with this over time. At some point, I will have to split Ryal Core out and then turn Ryal into its own OTP application.

**Comment by marjaimate** [comment]
FYI, example 'umbrella' Elixir app - https://github.com/wojtekmach/acme_bank/tree/master/apps

**Comment by BenMorganIO** [comment]
This is really good. I think we'll do this per mix package.
